address_book.rb

```diff
@@ -0,0 +1,293 @@
+require "yaml"
+require 'csv'
+require_relative 'person'
+require_relative 'address'
+class AddressBook
+
+   @@contacts=Array.new
+
+   def open
+     if File.exist?("contacts.yml")
+       @@contacts = YAML.load_file("contacts.yml")
+     end
+
+=begin
+     if File.exist?("contacts.csv")
+     @@contacts = CSV.read('contactBook.csv')
+   end #def (open)
+=end
+end
+   def save
+     File.open("contacts.yml", "w") do |file|
+       file.write(@@contacts.to_yaml)
+     end #file.open
+=begin
+     File.open("contactBook.csv", "w") do |file|
+       for i in 0..@@contacts.length-1
+       file.write(@@contacts[i])
+       file.write(@@contacts[i].getfirstName)
+       end
+     end #file.open
+=end
+   end
+
+ def isAlpha(word)
+   if /\d/.match( word)
+     return false
+   else
+     return true
+   end
+end
+
+
+   def isNum(num)
+     if /\d/.match( num)
+        return true
+     else
+      return false
+
+     end
+   end
+
+   def addContact
+     puts "Enter First Name"
+     fname=gets.chomp
+     while !( isAlpha(fname)) do
+       puts "Invalid name"
+       puts "Enter First name"
+       fname=gets.chomp
+     end
+
+     puts "Enter Last Name"
+     lname=gets.chomp
+
+     while !( isAlpha(lname)) do
+       puts "Invalid name"
+       puts "Enter Last name"
+       lname=gets.chomp
+     end
+
+     isExists=false
+     for i in 0..@@contacts.length-1
+       if fname==@@contacts[i].getfirstName and lname==@@contacts[i].getlastName
+
+         isExists=true
+         break
+       end
+     end
+     if isExists
+       puts "Contact already Exists"
+         displayContactByIndex(i)
+
+     else
+       puts "Enter phone number"
+       pnum=gets.chomp
+       while !(isNum(pnum)) do
+         puts "Invalid Phone Number"
+         puts "Enter Phone numebr"
+         pnum=gets.chomp
+       end
+
+       puts "Enter Address \n Street"
+       street=gets.chomp
+       while !( isAlpha(street) )do
+         puts "Invalid Street"
+         puts "Enter Street"
+         street=gets.chomp
+       end
+       puts"city"
+       city=gets.chomp
+       while !( isAlpha(city)) do
+         puts "Invalid city"
+         puts "Enter city"
+         city=gets.chomp
+       end
+       puts "state"
+       state=gets.chomp
+       while !( isAlpha(state)) do
+         puts "Invalid state"
+         puts "Enter state"
+         state=gets.chomp
+       end
+       puts "zip "
+       zip=gets.chomp
+       while !( isNum(zip)) do
+         puts "Invalid Zip "
+         puts "Enter Zip"
+         zip=gets.chomp
+       end
+   end
+       addr = Address.new(street,city,state,zip)
+       @@contacts.push(Person.new(fname,lname,pnum,addr))
+
+       end
+
```

Comment:
Can you please write a generic method which validates the entered value and also displays the appropriate message on invalid input, so that the repetitive lines in the "addContact" method from the above code can be avoided.